crypto/pkcs11key/pkcs11key.go

```diff
@@ -158,31 +129,67 @@ func (ps *PKCS11Key) Destroy() {
 	}
 }
 
-func (ps *PKCS11Key) openSession() (session pkcs11.SessionHandle, err error) {
+// Look up the token that contains the desired private key
+func (ps *PKCS11Key) openSession() (session pkcs11.SessionHandle, keyHandle pkcs11.ObjectHandle, err error) {
 	// Find slot by description
 	slots, err := ps.module.GetSlotList(true)
 	if err != nil {
 		return
 	}
 	for _, slot := range slots {
-		slotInfo, err := ps.module.GetSlotInfo(slot)
+		// If ps.slotDescription is provided, we will only check matching slots
+		slotInfo, err2 := ps.module.GetSlotInfo(slot)
+		if err2 != nil {
+			err = err2
+			fmt.Printf(">>> [%u] Couldn't get slot description\n", slot)
+			continue
+		}
+		if ps.slotDescription != "" && slotInfo.SlotDescription != ps.slotDescription {
+			fmt.Printf(">>> [%u] slotDescription didn't match [%s] != [%s]\n", slot, slotInfo.SlotDescription, ps.slotDescription)
+			continue
+		}
+
+		// Open session
+		session, err = ps.module.OpenSession(slot, pkcs11.CKF_SERIAL_SESSION)
 		if err != nil {
+			fmt.Printf(">>> [%u] Couldn't open session [%v]\n", slot, err)
 			continue
 		}
 
-		if slotInfo.SlotDescription == ps.slotDescription {
-			// Open session
-			session, err = ps.module.OpenSession(slot, pkcs11.CKF_SERIAL_SESSION)
-			if err != nil {
-				return session, err
-			}
+		// Login
+		if err = ps.module.Login(session, pkcs11.CKU_USER, ps.pin); err != nil {
+			fmt.Printf(">>> [%u] Couldn't log in with pin [%s] [%v]\n", slot, ps.pin, err)
+			ps.closeSession(session)
+			continue
+		}
 
-			// Login
-			if err = ps.module.Login(session, pkcs11.CKU_USER, ps.pin); err != nil {
-				return session, err
-			}
+		// See if the private key is present
+		template := []*pkcs11.Attribute{
+			pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PRIVATE_KEY),
+			pkcs11.NewAttribute(pkcs11.CKA_LABEL, ps.privateKeyLabel),
+		}
+		if err = ps.module.FindObjectsInit(session, template); err != nil {
+			fmt.Printf(">>> [%u] Couldn't FindObjectsInit\n", slot)
+			ps.closeSession(session)
```

Comment:
The repetition of `ps.closeSession()` suggests a refactor: If you put the inner part of this loop in its own function, you can follow `openSession` with `defer ps.closeSession()`, which will be shorter and more reliable.